src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -83,8 +83,8 @@ public TState GetRequiredState(DocumentId documentId)
         /// <summary>
         /// States ordered by <see cref="DocumentId"/>.
         /// </summary>
-        public IEnumerable<TState> States
-            => _map.Values;
```

Comment:
Add ImmutableSortedDictionary.Values to banned APIs?

Reply:
We could also make this a struct enumerator and keep the callers as they are.

Reply:
> Add ImmutableSortedDictionary.Values to banned APIs?

I wouldn't go this far. The `PerformanceSensitive` attribute is designed with the intent of statically validating this case where it matters.

> We could also make this a struct enumerator and keep the callers as they are.

This is a bit difficult. I added these struct enumerators when I wrote [`ImmutableSortedTreeDictionary<TKey, TValue>`](https://github.com/tunnelvisionlabs/dotnet-trees/blob/8c325603f17cc768d7e2cb86c18463f992cad013/TunnelVisionLabs.Collections.Trees/Immutable/ImmutableSortedTreeDictionary%602.cs#L46-L48) but it would be a binary breaking change to incorporate it back into System.Collections.Immutable. See dotnet/runtime#13927 for more details.

Reply:
Yeah, I understand we can't update  ImmutableSortedTreeDictionary, but we can either create an extension method `GetValues(ImmutableSortedTreeDictionary<K,V>)` or make specialized struct enumerator just for this property.